crud/src/main/java/com/redhat/lightblue/crud/TranslatorToJson.java

```diff
@@ -0,0 +1,212 @@
+import com.redhat.lightblue.metadata.ObjectArrayElement;
+import com.redhat.lightblue.metadata.ObjectField;
+import com.redhat.lightblue.metadata.ReferenceField;
+import com.redhat.lightblue.metadata.SimpleArrayElement;
+import com.redhat.lightblue.metadata.SimpleField;
+import com.redhat.lightblue.metadata.Type;
+import com.redhat.lightblue.util.Error;
+import com.redhat.lightblue.util.JsonDoc;
+import com.redhat.lightblue.util.Path;
+
+/**
+ * Defines a class that take a response from a datasource and translates it
+ * into something Lightblue can use.
+ *
+ * @author dcrissman
+ *
+ * @param <S> - The source type that this {@link TranslatorToJson} converts
+ * to json.
+ */
+public abstract class TranslatorToJson<S> {
+
+    private final JsonNodeFactory factory;
+    protected final EntityMetadata entityMetadata;
+
+    public TranslatorToJson(JsonNodeFactory factory, EntityMetadata entityMetadata){
+        this.factory = factory;
+        this.entityMetadata = entityMetadata;
+    }
+
+    protected JsonNode toJson(Type type, Object value){
+        return type.toJson(factory, value);
+    }
+
+    /**
+     * Translates the source to a {@link JsonDoc}.
+     * @param source - Object containing the source data.
+     * @return {@link JsonDoc}
+     */
+    public JsonDoc translate(S source){
+        Error.push("translating to json");
+        try{
+            FieldCursor cursor = entityMetadata.getFieldCursor();
+
+            if (cursor.firstChild()) {
+                ObjectNode node = factory.objectNode();
+
+                iterateOverNodes(source, node, cursor);
+
+                return new JsonDoc(node);
+            }
+        }
+        finally{
+            Error.pop();
+        }
+
+        //TODO: What to do in case of a null value here?
+        return null;
+    }
+
+    private void iterateOverNodes(Object value, ContainerNode<?> targetNode, FieldCursor cursor) {
+        do {
+            appendToJsonNode(value, targetNode, cursor);
+        } while(cursor.nextSibling());
+    }
+
+    protected void appendToJsonNode(Object value, ContainerNode<?> targetNode, FieldCursor cursor) {
+        FieldTreeNode field = cursor.getCurrentNode();
+
+        Error.push(field.getName());
+        try{
+            JsonNode newJsonNode = null;
+            Object newValue = null;
+
+            if (field instanceof ObjectField || field instanceof ObjectArrayElement) {
+                newJsonNode = translateToObjectNode(value, cursor);
+            }
+            else if((newValue = getValueFor(value, cursor.getCurrentPath())) != null){
+                if (field instanceof SimpleField) {
+                    newJsonNode = translate((SimpleField)field, newValue);
+                }
+                else if (field instanceof ArrayField){
+                    newJsonNode = translateToArrayNode((ArrayField) field, newValue, cursor);
+                }
+                else if (field instanceof ReferenceField) {
```

Comment:
Honestly, I wasn't sure how to handle ReferenceField(s), so for now the concrete implementations just have noop implementations.

> Or, don't even bother.  Once processed, and ReferenceField will be an ArrayField.

Are you suggesting that I just drop support for this field type? Or that I treat it like an ArrayField? ObjectArrayField maybe?